src/moonlink_backend/src/error.rs

```diff
@@ -1,89 +1,211 @@
+        let status = match &source {
+            #[cfg(feature = "metadata-postgres")]
+            MoonlinkMetadataStoreError::TokioPostgres(es) => es.status,
+            MoonlinkMetadataStoreError::PostgresRowCountError(es)
+            | MoonlinkMetadataStoreError::Sqlx(es)
+            | MoonlinkMetadataStoreError::SqliteRowCountError(es)
+            | MoonlinkMetadataStoreError::MetadataStoreFailedPrecondition(es)
+            | MoonlinkMetadataStoreError::SerdeJson(es)
+            | MoonlinkMetadataStoreError::TableIdNotFound(es)
+            | MoonlinkMetadataStoreError::ConfigFieldNotExist(es)
+            | MoonlinkMetadataStoreError::Io(es) => es.status,
+        };
+        Error::MoonlinkMetadataStoreError(ErrorStruct {
+            message: "Moonlink metadata store error".to_string(),
+            status,
+            source: Some(Arc::new(source.into())),
+            location: Some(Location::caller().to_string()),
+        })
     }
 }
 
 impl From<std::io::Error> for Error {
+    #[track_caller]
     fn from(source: std::io::Error) -> Self {
-        Error::Io {
-            source: Arc::new(source),
-        }
+        let status = match source.kind() {
+            std::io::ErrorKind::TimedOut
+            | std::io::ErrorKind::Interrupted
+            | std::io::ErrorKind::WouldBlock
+            | std::io::ErrorKind::ConnectionRefused
+            | std::io::ErrorKind::ConnectionAborted
+            | std::io::ErrorKind::ConnectionReset
+            | std::io::ErrorKind::BrokenPipe
+            | std::io::ErrorKind::NetworkDown
+            | std::io::ErrorKind::ResourceBusy
+            | std::io::ErrorKind::QuotaExceeded => ErrorStatus::Temporary,
+
+            _ => ErrorStatus::Permanent,
+        };
+
+        Error::Io(ErrorStruct {
+            message: format!("IO error: {source}"),
+            status,
+            source: Some(Arc::new(source.into())),
+            location: Some(Location::caller().to_string()),
+        })
     }
 }
 
 impl From<tokio::sync::watch::error::RecvError> for Error {
+    #[track_caller]
     fn from(source: tokio::sync::watch::error::RecvError) -> Self {
-        Error::TokioWatchRecvError { source }
+        Error::TokioWatchRecvError(ErrorStruct {
+            message: format!("Watch channel receive error: {source}"),
+            status: ErrorStatus::Permanent,
+            source: Some(Arc::new(source.into())),
+            location: Some(Location::caller().to_string()),
+        })
     }
 }
 
 impl From<serde_json::Error> for Error {
+    #[track_caller]
     fn from(source: serde_json::Error) -> Self {
-        Error::Json {
-            source: Arc::new(source),
-        }
+        let status = match source.classify() {
+            serde_json::error::Category::Io => ErrorStatus::Temporary,
+
+            _ => ErrorStatus::Permanent,
+        };
+
+        Error::Json(ErrorStruct {
+            message: format!("JSON serialization/deserialization error: {source}"),
```

Comment:
We are duplicating `source` information in both message and source field, prefer to store it only once.